### PR TITLE
Add master role label to server node at startup

### PR DESCRIPTION
**Pull request by galal-hussein**
https://github.com/rancher/k3s/issues/842

The PR now just adds the master role label on server startup

### CODE REVIEW

**Comment by yamt** [comment]
what's the purpose of this change?

**Comment by ibuildthecloud** [comment]
@yamt The node role labels must be assigned by a controller in 1.16.  The kubelets can't self assign anymore.

**Comment by ibuildthecloud** [comment]
@galal-hussein You you write a sentence or two how this controller works.  I don't fully understand the logic.  Regardless of the logic can you follow the pattern of: check if object needs to be updated, if not return, if so then DeepCopy, modify object, and call update.  So two important things.  Always DeepCopy before modifying an object and second, don't do an update unless it is need so that we reduce the number of writes.

File: pkg/noderole/controller.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+	coreclient "github.com/rancher/wrangler-api/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	core "k8s.io/api/core/v1"
+)
+
+const (
+	MasterRoleK3sLabelKey = "node-role.k3s.io/master"
+	MasterRoleLabelKey    = "node-role.kubernetes.io/master"
+	WorkerRoleLabelKey    = "node-role.kubernetes.io/worker"
```
Comment (yamt) [inline pkg/noderole/controller.go]: how about dropping the worker label? it doesn't seem to be a standard.

File: pkg/server/server.go [hunk 2]
```diff
@@ -357,3 +362,34 @@ func isSymlink(config string) bool {
...
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		if len(node.Labels) == 0 {
+			node.Labels = make(map[string]string)
+		}
+		if _, ok := node.Labels[MasterRoleLabelKey]; ok {
+			break
+		}
+		node.Labels[MasterRoleLabelKey] = "true"
```
Comment (ibuildthecloud) [inline pkg/server/server.go]: add DeepCopy

File: pkg/server/server.go [hunk 3]
```diff
@@ -357,3 +362,34 @@ func isSymlink(config string) bool {
...
+	for {
+		nodeName := os.Getenv("NODE_NAME")
+		nodeController := sc.Core.Core().V1().Node()
+		node, err := nodeController.Cache().Get(nodeName)
+		if err != nil {
+			logrus.Infof("Waiting for master node %s startup: %v", nodeName, err)
+			time.Sleep(1 * time.Second)
+			continue
+		}
+		if len(node.Labels) == 0 {
```
Comment (ibuildthecloud) [inline pkg/server/server.go]: check for nil, not len() == 0

File: pkg/server/server.go [hunk 4]
```diff
@@ -357,3 +362,34 @@ func isSymlink(config string) bool {
...
+			node.Labels = make(map[string]string)
+		}
+		if _, ok := node.Labels[MasterRoleLabelKey]; ok {
```
Comment (ibuildthecloud) [inline pkg/server/server.go]: check that `node.Labels[MasterRoleLabelKey] == "true"` to ensure the value is correct, not just set

File: pkg/server/server.go [hunk 5]
```diff
@@ -357,3 +362,34 @@ func isSymlink(config string) bool {
...
+			break
+		}
+		node.Labels[MasterRoleLabelKey] = "true"
+		_, err = nodeController.Update(node)
+		if err == nil {
+			logrus.Infof("master role labe has been set succesfully on node: %s", nodeName)
```
Comment (ibuildthecloud) [inline pkg/server/server.go]:
```suggestion
			logrus.Infof("master role label has been set successfully on node: %s", nodeName)
```

File: pkg/server/server.go (hunk 2 above)
Comment (galal-hussein) [reply pkg/server/server.go]: fixed

Comment (ibuildthecloud) [reply pkg/server/server.go]: Move deepcopy to line 383-ish

**Comment by ShylajaDevadiga** [comment]
Resolved. We have a master role label for master nodes with k3s v0.10.0-rc1

**Comment by huapox** [comment]
1.When server and agent run in spit mode `k3s server --disable-agent`, got these warnings:

```
server_1    | time="2019-10-24T11:24:22.415940766Z" level=info msg="Waiting for master node  startup: node \"\" not found"
server_1    | time="2019-10-24T11:24:23.416638995Z" level=info msg="Waiting for master node  startup: node \"\" not found"
server_1    | time="2019-10-24T11:24:24.417498743Z" level=info msg="Waiting for master node  startup: node \"\" not found"
server_1    | time="2019-10-24T11:24:25.417794752Z" level=info msg="Waiting for master node  startup: node \"\" not found"
server_1    | time="2019-10-24T11:24:26.418522541Z" level=info msg="Waiting for master node  startup: node \"\" not found"
server_1    | time="2019-10-24T11:24:27.419425228Z" level=info msg="Waiting for master node  startup: node \"\" not found"
server_1    | time="2019-10-24T11:24:28.420124957Z" level=info msg="Waiting for master node  startup: node \"\" not found"
server_1    | time="2019-10-24T11:24:29.420431712Z" level=info msg="Waiting for master node  startup: node \"\" not found"
```

2.Then the node joined the cluster:`kubectl label nodes node1 node-role.kubernetes.io/master=true`, to give the node a master role. but the warn is still printing in the server's log.

**Pull request merged**